clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java

```diff
@@ -737,16 +751,21 @@ public void handle(SyncGroupResponse syncResponse,
                     log.debug("SyncGroup failed because the group began another rebalance");
                     future.raise(error);
                 } else if (error == Errors.FENCED_INSTANCE_ID) {
-                    log.error("Received fatal exception: group.instance.id gets fenced");
+                    // for sync-group request, even if the generation has changed we would not expect the instance id
+                    // gets fenced, and hence we always treat this as a fatal error
+                    log.error("SyncGroup failed with {} due to group.instance.id {} gets fenced",
+                        sentGeneration, rebalanceConfig.groupInstanceId);
                     future.raise(error);
                 } else if (error == Errors.UNKNOWN_MEMBER_ID
                         || error == Errors.ILLEGAL_GENERATION) {
-                    log.debug("SyncGroup failed: {}", error.message());
-                    resetGenerationOnResponseError(ApiKeys.SYNC_GROUP, error);
+                    log.info("SyncGroup failed with {}: {}, would request re-join", sentGeneration, error.message());
+                    if (generationUnchanged())
```

Comment:
Hey @guozhangwang. 
Trying to understand if any of these changes have implications for the `WorkerCoordinator` too. 

What I've observed is that if the `WorkerCoordinator` fails to receive a valid `SyncGroup` request, it may miss several generations until it succeeds again. Is the branch here safe? `resetGenerationOnResponseError` will be called (and therefore rejoin will be requested) only if `generationUnchanged` returns true. Is there any risk that we suppress any useful retries to rejoin?

Reply:
The rationale is that if the generation has changed, it is either reset by the heartbeat thread in which case the generation is reset and rejoin is already requested, or it is changed by another join-group request; but since inside AbstractCoordinator we will only have one in-flight request at a given time the second scenario should not happen. So the only possibility is the heartbeat resetting.

Reply:
Makes sense. Thanks